heareval/predictions/task_predictions.py

```diff
@@ -997,49 +1123,77 @@ def print_scores(grid_point_results: List[GridPointResult]):
-        batch_size=conf["batch_size"],
-        in_memory=in_memory,
-    )
-    best_trainer = best_grid_point.trainer
-    # This hack is necessary because we use the best validation epoch to
-    # choose the event postprocessing
-    best_trainer.fit_loop.current_epoch = best_grid_point.epoch
+    # Train predictors for the remaining splits using the hyperparameters selected
+    # from the grid search.
+    split_grid_points = [best_grid_point]
+    for split in data_splits[1:]:
+        print(f"Training split: {split}")
+        grid_point_result = task_predictions_train(
+            embedding_path=embedding_path,
+            embedding_size=embedding_size,
+            metadata=metadata,
+            data_splits=split,
+            label_to_idx=label_to_idx,
+            nlabels=nlabels,
+            scores=scores,
+            conf=best_grid_point.conf,
+            gpus=gpus,
+            in_memory=in_memory,
+            deterministic=deterministic,
+        )
+        split_grid_points.append(grid_point_result)
+        print(
+            f"Split {split} validation score: ",
+            grid_point_result.validation_score,
+            embedding_path,
+        )
 
-    test_results = best_trainer.test(
-        ckpt_path=best_grid_point.model_path, test_dataloaders=test_dataloader
-    )
-    assert len(test_results) == 1, "Should have only one test dataloader"
-    test_results = test_results[0]
+    # Now test each of the trained models
+    test_results = {}
+    for i, split in enumerate(data_splits):
+        test_fold_str = "|".join(split["test"])
+        print(test_fold_str)
+        print(split_grid_points[i])
+        test_results[test_fold_str] = task_predictions_test(
+            embedding_path=embedding_path,
+            grid_point=split_grid_points[i],
+            metadata=metadata,
+            data_splits=split,
+            label_to_idx=label_to_idx,
+            nlabels=nlabels,
+            in_memory=in_memory,
+        )
 
-    test_results.update(
+        # Add model training values relevant to this split model
+        test_results[test_fold_str].update(
+            {
+                "validation_score": split_grid_points[i].validation_score,
+                "epoch": split_grid_points[i].epoch,
+                "time_in_min": split_grid_points[i].time_in_min,
+            }
+        )
+
+    # Update test results with values relevant for all split models
+    test_results[test_fold_str].update(
         {
-            "validation_score": best_grid_point.validation_score,
             "hparams": hparams_to_json(best_grid_point.hparams),
             "postprocessing": best_grid_point.postprocessing,
-            "epoch": best_grid_point.epoch,
-            "time_in_min": best_grid_point.time_in_min,
-            "score_mode": best_grid_point.score_mode,
+            "score_mode": split_grid_points[i].score_mode,
             "embedding_path": str(embedding_path),
         }
     )
+
```

Comment:
The function is getting a bit long, can we break it down?

Can we save the scores of the different folds?

Reply:
Did my best to pull some stuff out with out going crazy.

Reply:
I'm still kinda concerned tho, it's really long. Any way we can go further?

Reply:
#320